lib/saluki-event/src/metric/data.rs

```diff
@@ -0,0 +1,798 @@
+/// timestamp/value pair. This allows for holding multiple data points in a single metric.
+///
+/// ## Design
+///
+/// `MetricValues` is used to support two specific behaviors at the same time: holding values that may or may
+/// not have a timestamp, and holding multiple values attached to a single metric. This is a common representation of
+/// metrics, where a single metric may carry multiple values assigned to different buckets of time, such as the
+/// OpenTelemetry metrics data model. However, values may not yet be associated with a timestamp, such as when they
+/// arrive from applications that do not provide a timestamp.
+///
+/// Additionally, `MetricValues` only supports a homogenous set of values, which is either specified when creating an
+/// empty `MetricValues` (using [`MetricValues::from_kind`]) or when creating it from a single value (using
+/// [`MetricValues::from_value`].)
+///
+/// ## Timestamps
+///
+/// All metrics are associated with a timestamp value. The timestamp values themselves do not influence the behavior of
+/// `MetricValues`, and are only relevant towards merging and sorting behavior, but generally, a timestamp of zero is
+/// used to indicate no timestamp is (yet) associated with a value while non-zero timestamps are meant to indicate that
+/// a timestamp is already associated with a value and should be used.
+///
+/// ## Ordering
+///
+///
+///
+/// ## Creating and adding values
+///
+/// As noted above, `MetricValues` only holds a single kind of metric value at any given time, which is established
+/// creating `MetricValues` via [`MetricValues::from_kind`], [`MetricValues::from_value`],
+/// [`MetricValues::from_values`], or [`MetricValues::from_values_fallible`]. Once a `MetricValues` instance is created,
+/// further values can be added via [`MetricValues::merge_value`]. Even further, another `MetricValues` can be merged
+/// into an existing set of values via [`MetricValues::merge_values`].
+///
+/// For most use cases, adding _new_ values to `MetricValues` should prefer [`MetricValues::push_value`], which adds the
+/// value without any merging. This means that if a value already exists at the given timestamp, the new value will be
+/// be added to the set as a separate entry. This allows downstream code to be able to observe the original values and
+/// deal with any merging logic as needed.
+///
+/// ## Merging
+///
+/// In many cases, callers will either want to merge values with the same timestamp, whether merging single values or
+/// merging another set of values. Both [`MetricValues::merge_value`] and [`MetricValues::merge_values`] support this
+/// behavior by merging the incoming value(s) with any existing value(s) with the same timestamp. If no value exists at
+/// the given timestamp, new entries will be added.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct MetricValues {
+    kind: MetricKind,
+    values: SmallVec<[(u64, MetricValue); 2]>,
+}
```

Comment:
Similar to my previous comment, my biggest question here is whether we want to have these two axes mixed within a single collection (timestamped vs not and potentially values with different kinds). It kind of seems like we're allowing a lot of potentially possibilities to be represented and I'm not sure how many of are really valid/useful. If we can split these possibilities out into more structured forms, the code that handles them can likely get quite a bit simpler.

Reply:
Yeah, I think to pull it out a little bit, there's a bunch of very real scenarios in terms of what we see coming into the DSD source:

- single metric value w/o timestamp (arguably the most common scenario)
- multiple metric values w/o timestamp (I don't have the data on how common this is, but I'd guess much much less common than single metric value)
- single metric value w/ timestamp (probably becoming more common.. maybe more common than multiple-value-no-timestamp?)
- multiple metric values w/ timestamp (technically possible, but not likely since having a timestamp implies aggregation, which implies a single value)

So our most _common_ scenario is handling metric values without a timestamp, with a very real possibility of handling multiple of this. After that, it's single timestamped value.

After that, then there's what we need to handle coming out of the aggregator phase, which is single or multiple values with timestamps: they _always_ have a timestamp, and it's just as likely to have one timestamped value as it is to have two (the two is just a factor of the flush interval, and will likely only ever be a max of two... but I'm not sure it's safe to use that as an invariant or if it buys us anything by doing so).

Kind of stream of consciousness spewing here... but yeah, I _do_ think this design is only as complex as it needs to be to support the relevant scenarios with a single data type, but related to your comment above, I agree that it's not the most optimal design and we could definitely think about trying to bifurcate some of this a little more.

I'll give some more thought to that approach.